CanvasRenderingContext2D.fontStretch - make live example work (#20794)
mdn/content: files/en-us/web/api/canvasrenderingcontext2d/fontstretch/index.md

### JavaScript

> First we get the canvas declared in the HTML file and use it to get the "2d context" that will later be used for drawing text.

Reviewer: Not sure if this suggestion is best, but not a fan of "2d context". Could probably also just be there without any formatting.

```suggestion
First we get the canvas declared in the HTML file and use it to get the `2d` context that will later be used for drawing text.
```

Author: Yes. A bit ugly. I think I'll modify it like this:
```suggestion
First we get the canvas declared in the HTML file and use it to get the `CanvasRenderingContext2D` that will later be used for drawing text.
```

```js
const canvas = document.getElementById("canvas");
const ctx = canvas.getContext("2d");
```

> The next step in the example is to load a variable font that can be varied in the width axis.

Author: The document is about using fontStretch not about using Google Fonts and how you find and use them. Given how hard it was to find out how to do this I did agonize a bit, but I think this is the right thing as it might be seen as promoting one particular font technology.

FYI only though, in case you have to do this:
1. YOu can filter on just variable fonts from the front page
   ![image](https://user-images.githubusercontent.com/5368500/191185404-248c0eb2-c6e0-4317-9387-4e471f57c11d.png)
  But these can vary on any axis. You want a font that varies on width like inconsolata:
  ![image](https://user-images.githubusercontent.com/5368500/191185706-15b5de21-7054-4972-b965-dbb133022291.png)
2. There is a way to list them with their axes, but I can't find it.
3. Then for this API you can't just pass in the URL for the page, you have to construct a URL that says what you want. THat gives you a bunch of font files, and you then pick the one you want. If you click this link you will see what I mean https://fonts.googleapis.com/css2?family=Inconsolata:wdth@50..200 - this is for the inconsolata file that has the right values for the stretch fonts.

> The next step in the example is to load a variable font that can be varied in the width axis.

Author: An easier way to find all the variable fonts with axis: https://fonts.google.com/variablefonts

> The next step in the example is to load a variable font that can be varied in the width axis.

Reviewer 2: >The document is about using fontStretch not about using Google Fonts and how you find and use them

Agreed, I think what you've written is good.

This is needed because `fontStretch` can only stretch a font that contains information about how glyphs are drawn when stretched — otherwise text will be drawn using the closest available font stretch value for the font, which will often be the normal width.

> In this case we use [`FontFace`](/en-US/docs/Web/API/FontFace) to define a font face for the [Inconsolata](https://fonts.google.com/specimen/Inconsolata/tester) Google Font, which can be stretched from 50% to 200% (the full range).

Reviewer 2: What does "the full range" mean?

Author: I mean that the API allows for values from 0 to 200% and maps the keywords to them.

```suggestion
In this case we use [`FontFace`](/en-US/docs/Web/API/FontFace) to define a font face for the [Inconsolata](https://fonts.google.com/specimen/Inconsolata/tester) Google Font, which supports font widths from 50% to 200% (allowing us to demonstrate `fontStretch` values from `ultra-condensed` to `ultra-expanded`).
```

We then add this to the document's [`FontFaceSet`](/en-US/docs/Web/API/FontFaceSet) ([`document.fonts`](/en-US/docs/Web/API/Document/fonts)) so that it can be used for drawing.

```js
const fontFile = new FontFace(
  "Inconsolata",
  'url(https://fonts.gstatic.com/s/inconsolata/v31/QlddNThLqRwH-OJ1UHjlKENVzlm-WkL3GZQmAwPyya15.woff2) format("woff2")',
  { stretch: "50% 200%" }
);

document.fonts.add(fontFile);
```

> The code below then calls [`FontFaceSet.load()`](/en-US/docs/Web/API/FontFaceSet/load) to fetch and load the google font.

Reviewer: ```suggestion
The code below then calls [`FontFaceSet.load()`](/en-US/docs/Web/API/FontFaceSet/load) to fetch and load the Google font.
```

Note that this call sets the size of the font that is needed, and returns a promise that resolves when the font has been loaded.

We then assign the font face we downloaded to the context, and use the context to draw text to the canvas at each of the keyword stretch levels.

> Note that again the size of the desired font is specified.

Author: ```suggestion
Note that again the size of the desired font is specified (this does not have to match the loaded font size).
```

> This does not have to match the loaded font size; but may not render as well if there is a mismatch.

Author: I know that you can load a font in one size and draw in another. 

My _understanding_ is that this is because the font engine can scale the size font it has downloaded. So you will get best result if they loaded and drawn fonts are the same, but it will still work. However I am not 100% certain on this, so if you know differently  ....

Reviewer: I _think_ you are right.

Reviewer 2: I tested with 

```js
document.fonts.load("12px Inconsolata").then(
  () => {
    ctx.font = "150px 'Inconsolata'";
    // ...
```

versus

```js
document.fonts.load("150px Inconsolata").then(
  () => {
    ctx.font = "12px 'Inconsolata'";
    // ...
```

It doesn't have any visible side-effects of upscaling / downscaling, at least.

Author: OK, I am deleting this and adding the comment above: https://github.com/mdn/content/pull/20794/files#r974955540

Essentially saying it doesn't have to be the same font size, but not explaining why.
That allows us to merge this as "not wrong" and address reasons later.

```suggestion
```

@Elchi3 Do you know someone who could explain this/confirm?

It might well be that it works for this particular type of font, but it might not work in another case. That's why I used the weasel words "but _may not_ render as well if there is a mismatch."

Reviewer: Trying my luck and pinging @fserb here. Fernando, do loaded font size and ctx specified font sizes have to match? What if they don't?

```
document.fonts.load("30px Inconsolata").then(
  () => {
    ctx.font = "30px 'Inconsolata'";
```

Author: @fserb ^^^ Any advice you can give would be much appreciated. I'll merge this next week if we don't get a response.

```js
document.fonts.load("30px Inconsolata").then(
  () => {
    ctx.font = "30px 'Inconsolata'";

>     // Default (normal)

Author: Leaving this one - it is not obvious otherwise this is "normal"

    ctx.fillText(`Hello world (default: ${ctx.fontStretch})`, 5, 20);

>     // Font stretch: ultra-condensed
>     ctx.fontStretch = "60%";
>     //ctx.fontStretch = 'ultra-condensed';
>     ctx.fillText(`Hello world (${ctx.fontStretch})`, 5, 50);

Author: ```suggestion
    ctx.fontStretch = "ultra-condensed";
    ctx.fillText(`Hello world (${ctx.fontStretch})`, 5, 50);
```

>     // Font stretch: extra-condensed

Reviewer: I think our coding standards say that we shouldn't have comments that repeat what the code already says. I think we can make this a lot shorter and remove all these comments.

    ctx.fontStretch = "extra-condensed";
    ctx.fillText(`Hello world (${ctx.fontStretch})`, 5, 80);

>     // Font stretch: condensed
>     ctx.fontStretch = "condensed";
>     ctx.fillText(`Hello world (${ctx.fontStretch})`, 5, 110);
> 
>     // Font stretch: semi-condensed
>     ctx.fontStretch = "semi-condensed";
>     ctx.fillText(`Hello world (${ctx.fontStretch})`, 5, 140);
> 
>     // Font stretch: extra-condensed
>     ctx.fontStretch = "extra-condensed";
>     ctx.fillText(`Hello world (${ctx.fontStretch})`, 5, 170);
> 
>     // Font stretch: semi-expanded
>     ctx.fontStretch = "semi-expanded";
>     ctx.fillText(`Hello world (${ctx.fontStretch})`, 5, 200);
> 
>     // Font stretch: expanded
>     ctx.fontStretch = "expanded";
>     ctx.fillText(`Hello world (${ctx.fontStretch})`, 5, 230);
> 
>     // Font stretch: extra-expanded
>     ctx.fontStretch = "extra-expanded";
>     ctx.fillText(`Hello world (${ctx.fontStretch})`, 5, 260);
> 
>     // Font stretch: ultra-expanded
>     ctx.fontStretch = "ultra-expanded";
>     ctx.fillText(`Hello world (${ctx.fontStretch})`, 5, 290);

Author: ```suggestion
    ctx.fontStretch = "condensed";
    ctx.fillText(`Hello world (${ctx.fontStretch})`, 5, 110);

    ctx.fontStretch = "semi-condensed";
    ctx.fillText(`Hello world (${ctx.fontStretch})`, 5, 140);

    ctx.fontStretch = "extra-condensed";
    ctx.fillText(`Hello world (${ctx.fontStretch})`, 5, 170);

    ctx.fontStretch = "semi-expanded";
    ctx.fillText(`Hello world (${ctx.fontStretch})`, 5, 200);

    ctx.fontStretch = "expanded";
    ctx.fillText(`Hello world (${ctx.fontStretch})`, 5, 230);

    ctx.fontStretch = "extra-expanded";
    ctx.fillText(`Hello world (${ctx.fontStretch})`, 5, 260);

    ctx.fontStretch = "ultra-expanded";
    ctx.fillText(`Hello world (${ctx.fontStretch})`, 5, 290);
```
